lib/Sema/CSSolver.cpp

```diff
@@ -2431,6 +2426,251 @@ determineBestBindings(ConstraintSystem &CS) {
+
+    float componentScore = 0.0;
+    for (auto typeVar : typeVars)
+      componentScore += scoreTypeVariable(cs, typeVar);
+
+    // Average component score based on the number of type variables.
+    return componentScore / typeVars.size();
+  };
+
+  auto fnType = type->getAs<AnyFunctionType>();
+  if (!fnType)
+    return scoreComponent(type);
+
+  unsigned components = 1;
+  // Result could either be a type variable or concrete type or a function.
+  score += scoreComponent(fnType->getResult());
+
+  // Input type could be only ParenType or TupleType.
+  auto inputType = fnType->getInput();
+  if (auto paren = dyn_cast<ParenType>(inputType.getPointer())) {
+    score += scoreComponent(paren->getUnderlyingType());
+    ++components;
+  } else if (auto tuple = inputType->getAs<TupleType>()) {
+    for (auto &element : tuple->getElements()) {
+      score += scoreComponent(element.getType());
+      ++components;
+    }
+  }
+
+  // Average score based number of the components in the type
+  // that makes sure that we are fair to the functions with
+  // different number of argument/result types.
+  return score / components;
+}
+
+Optional<Constraint *> ConstraintSystem::selectDisjunction(
+    llvm::SmallVectorImpl<Constraint *> &disjunctions) {
+  if (disjunctions.empty())
+    return None;
+
+  if (disjunctions.size() == 1) {
+    auto disjunction = disjunctions[0];
+    // If there was only one disjunction available and it doesn't
+    // have any choices enabled, that means that system won't have
+    // solution and this disjunction can't be picked.
+    if (disjunction->countActiveNestedConstraints() == 0)
+      return None;
+
+    return disjunction;
+  }
+
+  Constraint *bestDisjunction = nullptr;
+  float bestScore = 0.0;
+
+  auto getDisjunctionId = [&](Constraint *disjunction) -> unsigned {
+    assert(disjunction->getKind() == ConstraintKind::Disjunction);
+    auto *const choice = disjunction->getNestedConstraints().front();
+    auto type = choice->getFirstType();
+
+    if (auto typeVar = type->getAs<TypeVariableType>())
+      return typeVar->getID();
+
+    if (auto fnType = type->getAs<FunctionType>()) {
+      auto resultType = fnType->getResult();
+      if (auto typeVar = resultType->getAs<TypeVariableType>())
+        return typeVar->getID();
+    }
+
+    // If it's neither a type variable (bind overload) or function
+    // with type variable as a result type let's push it to the front
+    // of the list with all else equal.
+    return 0;
+  };
+
+  auto evaluateDisjunction = [&](Constraint *contender, float score) {
+    // Score has to be strictly greather than best because that would
+    // allow us to apply disjunctions in their semantic order, which gives
+    // better changes of avoiding of the the obviously wrong choices with
+    // everything else being equal.
+    if (score > bestScore ||
+        (score == bestScore &&
+         getDisjunctionId(bestDisjunction) > getDisjunctionId(contender))) {
```

Comment:
Rather than doing a bit of work to dig out type variable IDs to compare (which isn't always going to give an ordering), why not keep track of the index of the best option within the disjunction? It'll be more predictable and eliminate the need for `getDisjunctionId`.